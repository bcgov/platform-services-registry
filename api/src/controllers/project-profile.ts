//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import { getAuthorization } from '../libs/authorization';
import { getDefaultCluster } from '../libs/profile';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserIdOrEmail(user.id, user.email);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const projectDetails = await ProfileModel.findById(Number(profileId));

    const isAuthorized = getAuthorization(profileId, user, projectDetails);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    res.status(200).json(projectDetails);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix OR primary cluster name (current default cluster).
  // If it exists, this overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
    primaryClusterName: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();
    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const defaultCluster = await getDefaultCluster();
    if (!defaultCluster) {
      throw errorWithCode(500, 'Unable to set primary cluster id based on default cluster');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix, primaryClusterName: defaultCluster.name });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
    migratingLicenseplate,
  } = body;

  try {
    const currentProjectDetails = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      busOrgId,
      prioritySystem,
      userId: currentProjectDetails.userId,
      namespacePrefix: currentProjectDetails.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
      primaryClusterName: currentProjectDetails.primaryClusterName,
      migratingLicenseplate,
    };

    const isAuthorized = getAuthorization(profileId, user, currentProjectDetails);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const isAuthorized = getAuthorization(profileId, user);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
