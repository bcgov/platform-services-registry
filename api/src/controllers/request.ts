//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { HumanActionType, RequestType } from '../db/model/request';
import { AuthenticatedUser } from '../libs/authmware';
import { fulfillRequest } from '../libs/fulfillment';
import { MessageType, sendProvisioningMessage } from '../libs/messaging';
import { archiveProjectSet } from '../libs/profile';
import shared from '../libs/shared';

const dm = new DataManager(shared.pgPool);

export const fetchHumanActionRequests = async (
  { query }: { query: any }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  const { filter } = query;

  try {
    const filterValue = true;
    // Step 1. fetch all requests with matching filter
    const requests = await RequestModel.findActiveByFilter(filter, filterValue);
    // Step 2. return JSON of project Ids
    res.status(200).json(requests);
  } catch (err) {
    const message = `Unable to fetch requests requiring human action`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateRequestHumanAction = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  const { requestId } = params;
  const { type, comment } = body;

  try {
    // Step 1. fetch Request
    const request = await RequestModel.findById(requestId);

    // Step 2. create human_action record
    await RequestModel.createHumanAction({
      requestId: request.id,
      type,
      comment,
      userId: user.id,
    });

    // Step 3.a. If approved: fulfillRequest functionality => create bot_message, send NATS message
    // Step 3.b. if rejected: updateRejectProject => archive ProjectSet, Email PO/TC with comment, complete request;
    if (type === HumanActionType.Approve) {
      await fulfillRequest(request);
      await RequestModel.receivedHumanAction(requestId);
      return res.status(204).end();
    }

    logger.info(`Sending CHES message Project Request Rejected for ${request.profileId}`);
    await sendProvisioningMessage(request.profileId, MessageType.RequestRejected);
    logger.info(`CHES message sent for ${request.profileId}`);

    if ( request.type === RequestType.Create) {
      await archiveProjectSet(request.profileId)
    }

    await RequestModel.updateCompletionStatus(requestId);
    return res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable to update request ${requestId} with human action`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
