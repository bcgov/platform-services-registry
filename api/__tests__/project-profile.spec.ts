//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import fs from 'fs';
import { camelCase } from 'lodash';
import path from 'path';
import { Pool } from 'pg';
import { archiveProjectProfile, createProjectProfile, fetchAllProjectProfiles, fetchProjectProfile, uniqueNamespacePrefix, updateProjectProfile } from '../src/controllers/project-profile';
import ProfileModel from '../src/db/model/profile';
import { AccessFlag } from '../src/libs/authorization';
import { requestProjectProfileEdit } from '../src/libs/request';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-profiles.json');
const selectProfiles = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/insert-profile.json');
const insertProfile = JSON.parse(fs.readFileSync(p1, 'utf8'));

const p2 = path.join(__dirname, 'fixtures/get-authenticated-user.json');
const authenticatedUser = JSON.parse(fs.readFileSync(p2, 'utf8'));

const p3 = path.join(__dirname, 'fixtures/select-default-cluster.json');
const selectCluster = JSON.parse(fs.readFileSync(p3, 'utf8'));

const p4 = path.join(__dirname, 'fixtures/select-profile.json');
const selectProfile = JSON.parse(fs.readFileSync(p4, 'utf8'));

const client = new Pool().connect();


jest.mock('../src/libs/request', () => ({
  requestProjectProfileEdit: jest.fn().mockResolvedValue({id: 2}),
}));

jest.mock('../src/db/utils', () => ({
  generateNamespacePrefix: jest.fn().mockReturnValue('c8c7e6'),
  transformKeysToCamelCase: jest.fn().mockImplementation(data => {
    const obj = {};
    Object.keys(data).forEach(key => {
      obj[camelCase(key)] = data[key];
    });

    return obj;
  }),
}));

jest.mock('../src/libs/fulfillment', () => {
  return {
    fulfillRequest: jest.fn(),
  };
});

describe('Project-profile event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();

    ex = new FauxExpress();
  });

  it('A project-profile is created on a given prod-ready cluster', async () => {
    const req = {
      body: insertProfile,
      user: {
        id: 1,
      },
    };
    const addon = {
      id: 9,
      createdAt: new Date(),
      updatedAt: new Date(),
    };

    client.query.mockReturnValueOnce({ rows: selectCluster });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ ...insertProfile, ...addon }] });

    // @ts-ignore
    await createProjectProfile(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot({
      // id: expect.any(Number),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('A project-profile is created on a given test cluster', async () => {
    const req = {
      body: { ...insertProfile, primaryClusterName: 'clab' },
      user: { ...authenticatedUser, accessFlags: [AccessFlag.ProvisionOnTestCluster,] },
    };
    const addon = {
      id: 9,
      createdAt: new Date(),
      updatedAt: new Date(),
    };

    const testCluster = {
      name: 'clab',
      isProd: false
    };

    client.query.mockReturnValueOnce({ rows: [testCluster] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ ...insertProfile, ...addon }] });

    // @ts-ignore
    await createProjectProfile(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot({
      // id: expect.any(Number),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('A project-profile fails to create', async () => {
    const req = {
      body: insertProfile,
    };

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(createProjectProfile(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });

  it('All project-profiles are returned', async () => {
    const req = {
      user: authenticatedUser,
    };

    client.query.mockReturnValueOnce({ rows: selectProfiles });

    await fetchAllProjectProfiles(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('All project-profiles are returned to an administrator', async () => {
    const req = {
      user: { ...authenticatedUser, accessFlags: [AccessFlag.EditAll,] },
    };

    const findAll = ProfileModel.prototype.findAll = jest.fn();
    const findProfilesByUserId = ProfileModel.prototype.findProfilesByUserIdOrEmail = jest.fn();

    await fetchAllProjectProfiles(req, ex.res);

    expect(findAll).toHaveBeenCalledTimes(1);
    expect(findProfilesByUserId).toHaveBeenCalledTimes(0);
  });

  it('All project-profiles are returned to non-administrator',
    async () => {
      const req = {
        user: authenticatedUser,
      };

      const findAll = ProfileModel.prototype.findAll = jest.fn();
      const findProfilesByUserId = ProfileModel.prototype.findProfilesByUserIdOrEmail = jest.fn();

      await fetchAllProjectProfiles(req, ex.res);

      expect(findAll).toHaveBeenCalledTimes(0);
      expect(findProfilesByUserId).toHaveBeenCalledTimes(1);
      expect(findProfilesByUserId.mock.calls[0][0]).toBe(req.user.id);
    });

  it('Fetch all project-profiles should throw', async () => {
    const req = {};
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(fetchAllProjectProfiles(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('A single project-profile is returned', async () => {
    const req = {
      params: { profileId: 4 },
    };
    client.query.mockReturnValueOnce({ rows: selectProfile });

    await fetchProjectProfile(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch single project-profiles should throw', async () => {
    const req = {
      params: { profileId: 4 },
    };
    client.query.mockImplementation(() => { throw new Error() });

    await expect(fetchProjectProfile(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('Update a project-profile with non provisioner-related changes', async () => {
    const body = {
      id: 4,
      name: "Project X",
      description: "This is a cool project.",
      criticalSystem: false,
      prioritySystem: false,
      busOrgId: "CITZ",
      notificationEmail: true,
      notificationSms: true,
    };
    const req = {
      params: { profileId: 4 },
      body,
      user: authenticatedUser,
    };

    client.query.mockReturnValueOnce({ rows: selectProfile });
    client.query.mockReturnValueOnce({ rows: [] });  


    const update = ProfileModel.prototype.update = jest.fn();

    await updateProjectProfile(req, ex.res);

    expect(update).toHaveBeenCalledTimes(1);
    expect(requestProjectProfileEdit).toHaveBeenCalledTimes(1);
    expect(ex.res.statusCode).toBe(204);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('request project-profile edit with provisioner-related name changes', async () => {
    const body = {
      id: 4,
      name: "Project X ATTEMPTED NAME CHANGE",
      description: "This is a cool project.",
      criticalSystem: false,
      prioritySystem: false,
      busOrgId: "CITZ",
      notificationEmail: true,
      notificationSms: true,
    };
    const req = {
      params: { profileId: 4 },
      body,
      user: authenticatedUser,
    };

    client.query.mockReturnValueOnce({ rows: selectProfile });

    const update = ProfileModel.prototype.update = jest.fn();

    await updateProjectProfile(req, ex.res);

    expect(update).toHaveBeenCalledTimes(0);
    expect(requestProjectProfileEdit).toHaveBeenCalledTimes(1);
    expect(ex.res.statusCode).toBe(202);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('request project-profile edit with provisioner-related description changes', async () => {
    const body = {
      id: 4,
      name: "Project X",
      description: "This is a cool project with modified description",
      criticalSystem: false,
      prioritySystem: false,
      busOrgId: "CITZ",
    };
    const req = {
      params: { profileId: 4 },
      body,
      user: authenticatedUser,
    };

    client.query.mockReturnValueOnce({ rows: selectProfile }); 

    const update = ProfileModel.prototype.update = jest.fn();

    await updateProjectProfile(req, ex.res);

    expect(update).toHaveBeenCalledTimes(0);
    expect(requestProjectProfileEdit).toHaveBeenCalledTimes(1);
    expect(ex.res.statusCode).toBe(202);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('A project-profiles fails to update', async () => {
    const aBody = {
      id: 4,
      ...insertProfile,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    };
    const req = {
      params: { profileId: 4 },
      body: aBody,
      user: authenticatedUser,
    }

    client.query.mockImplementation(() => { throw new Error() });

    await expect(updateProjectProfile(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });

  it('A project-profiles is archived', async () => {
    const aBody = {
      id: 4,
      ...insertProfile,
      createdAt: Date.now(),
      updatedAt: Date.now(),
      userId: 1,
    };
    const req = {
      params: { profileId: 4 },
    }

    client.query.mockReturnValue({ rows: [aBody] });

    await archiveProjectProfile(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.json).not.toBeCalled();
    expect(ex.res.end).toBeCalled();
  });

  it('A project-profiles fails to archive', async () => {
    const req = {
      params: { profileId: 4 },
    }

    client.query.mockImplementation(() => { throw new Error() });

    await expect(archiveProjectProfile(req, ex.res)).rejects.toThrow();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.res.json).not.toBeCalled();
    expect(ex.res.end).not.toBeCalled();
  });


  it('Unique project-profiles names are derived', async () => {
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });

    const result = await uniqueNamespacePrefix();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });
});
