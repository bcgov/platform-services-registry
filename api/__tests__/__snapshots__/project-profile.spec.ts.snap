// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project-profile event handlers A project fails to archive 1`] = `
Array [
  Array [
    Object {
      "text": "
        UPDATE profile
          SET
            archived = true
          WHERE id = 4
          RETURNING *;
      ",
    },
  ],
]
`;

exports[`Project-profile event handlers A project fails to archive 2`] = `200`;

exports[`Project-profile event handlers A project is archived 1`] = `
Array [
  Array [
    Object {
      "text": "
        UPDATE profile
          SET
            archived = true
          WHERE id = 4
          RETURNING *;
      ",
    },
  ],
]
`;

exports[`Project-profile event handlers A project is archived 2`] = `204`;

exports[`Project-profile event handlers A project is archived 3`] = `undefined`;

exports[`Project-profile event handlers A project is created 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT COUNT(*) FROM profile
          WHERE namespace_prefix = $1;",
      "values": Array [
        "c8c7e6",
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT COUNT(*) FROM profile
          WHERE namespace_prefix = $1;",
      "values": Array [
        "c8c7e6",
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT COUNT(*) FROM profile
          WHERE namespace_prefix = $1;",
      "values": Array [
        "c8c7e6",
      ],
    },
  ],
  Array [
    Object {
      "text": "
      SELECT * FROM ref_cluster
        WHERE archived = false;
      ",
    },
  ],
  Array [
    Object {
      "text": "
        INSERT INTO profile
          (name, description, bus_org_id, priority_system,
            critical_system, user_id, namespace_prefix,
            notification_email, notification_sms, notification_ms_teams,
            payment_bambora, payment_pay_bc, file_transfer, file_storage,
            geo_mapping_web, geo_mapping_location, scheduling_calendar,
            scheduling_appointments, idm_site_minder,
            idm_keycloak, idm_active_dir, other, primary_cluster_name,
            migrating_licenseplate)
          VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12,
            $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24) RETURNING *;",
      "values": Array [
        "Health Gateway",
        "Patient prescription history website",
        "HLTH",
        false,
        false,
        1,
        "c8c7e6",
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        "This is a test",
        "silver",
        undefined,
      ],
    },
  ],
]
`;

exports[`Project-profile event handlers A project is created 2`] = `200`;

exports[`Project-profile event handlers A project is created 3`] = `
Object {
  "busOrgId": "HLTH",
  "categoryId": 1,
  "createdAt": Any<Date>,
  "criticalSystem": false,
  "description": "Patient prescription history website",
  "fileStorage": false,
  "fileTransfer": false,
  "geoMappingLocation": false,
  "geoMappingWeb": false,
  "id": 9,
  "idmActiveDir": false,
  "idmKeyCloak": false,
  "idmSiteMinder": false,
  "name": "Health Gateway",
  "notificationEmail": false,
  "notificationMsTeams": false,
  "notificationSms": false,
  "other": "This is a test",
  "paymentBambora": false,
  "paymentPayBc": false,
  "primaryClusterName": "silver",
  "prioritySystem": false,
  "schedulingAppointments": false,
  "schedulingCalendar": false,
  "updatedAt": Any<Date>,
}
`;

exports[`Project-profile event handlers A project is updated 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        UPDATE profile
          SET
            name = $1, description = $2, bus_org_id = $3,
            priority_system = $4, notification_email = $5, notification_sms = $6,
            notification_ms_teams = $7, payment_bambora = $8, payment_pay_bc = $9,
            file_transfer = $10, file_storage = $11, geo_mapping_web = $12,
            geo_mapping_location = $13, scheduling_calendar = $14, scheduling_appointments = $15,
            idm_site_minder = $16, idm_keycloak = $17,
            idm_active_dir = $18, other = $19, primary_cluster_name = $20, migrating_licenseplate = $21
          WHERE id = 4
          RETURNING *;",
      "values": Array [
        "Health Gateway",
        "Patient prescription history website",
        "HLTH",
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        "This is a test",
        "silver",
        undefined,
      ],
    },
  ],
]
`;

exports[`Project-profile event handlers A project is updated 2`] = `200`;

exports[`Project-profile event handlers A project is updated 3`] = `
Object {
  "busOrgId": "HLTH",
  "categoryId": 1,
  "createdAt": "2020-05-19T20:02:54.561Z",
  "criticalSystem": false,
  "description": "Patient prescription history website",
  "fileStorage": false,
  "fileTransfer": false,
  "geoMappingLocation": false,
  "geoMappingWeb": false,
  "id": 4,
  "idmActiveDir": false,
  "idmKeyCloak": false,
  "idmSiteMinder": false,
  "name": "Health Gateway",
  "notificationEmail": false,
  "notificationMsTeams": false,
  "notificationSms": false,
  "other": "This is a test",
  "paymentBambora": false,
  "paymentPayBc": false,
  "primaryClusterName": "silver",
  "prioritySystem": false,
  "schedulingAppointments": false,
  "schedulingCalendar": false,
  "updateAt": "2020-05-19T20:02:54.561Z",
  "userId": 1,
}
`;

exports[`Project-profile event handlers A single profile is returned 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Project-profile event handlers A single profile is returned 2`] = `200`;

exports[`Project-profile event handlers A single profile is returned 3`] = `
Object {
  "archived": false,
  "busOrgId": "CITZ",
  "createdAt": "2020-09-10T18:14:13.304Z",
  "criticalSystem": false,
  "description": "This is a cool project.",
  "id": 4,
  "name": "Project X",
  "namespacePrefix": "4ea35c",
  "primaryClusterName": "silver",
  "prioritySystem": false,
  "updatedAt": "2020-09-10T18:14:13.304Z",
  "userId": 1,
}
`;

exports[`Project-profile event handlers All profiles are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
      SELECT DISTINCT(profile.id), profile.* FROM profile
      JOIN profile_contact ON profile_contact.profile_id = profile.id
      JOIN contact ON contact.id = profile_contact.contact_id
      WHERE (profile.user_id = $1 OR contact.email = $2) AND profile.archived = false;",
      "values": Array [
        1,
        "jane@example.com",
      ],
    },
  ],
]
`;

exports[`Project-profile event handlers All profiles are returned 2`] = `200`;

exports[`Project-profile event handlers All profiles are returned 3`] = `
Array [
  Object {
    "archived": false,
    "busOrgId": "CITZ",
    "createdAt": "2020-04-28T00:00:00.000Z",
    "criticalSystem": false,
    "description": null,
    "id": 1,
    "name": "Health Gateway",
    "namespacePrefix": "elf077",
    "primaryClusterName": "silver",
    "prioritySystem": false,
    "updatedAt": "2020-04-28T00:00:00.000Z",
    "userId": 1,
  },
  Object {
    "archived": false,
    "busOrgId": "CITZ",
    "createdAt": "2020-04-28T00:00:00.000Z",
    "criticalSystem": true,
    "description": "Hello World",
    "id": 2,
    "name": "EPIC",
    "namespacePrefix": "101ed4",
    "primaryClusterName": "silver",
    "prioritySystem": false,
    "updatedAt": "2020-04-28T00:00:00.000Z",
    "userId": 1,
  },
  Object {
    "archived": false,
    "busOrgId": "CITZ",
    "createdAt": "2020-04-28T00:00:00.000Z",
    "criticalSystem": false,
    "description": "This is some description",
    "id": 3,
    "name": "Mines Digital Services",
    "namespacePrefix": "ae11ea",
    "primaryClusterName": "silver",
    "prioritySystem": false,
    "updatedAt": "2020-04-28T00:00:00.000Z",
    "userId": 1,
  },
  Object {
    "archived": false,
    "busOrgId": "CITZ",
    "createdAt": "2020-09-10T18:14:13.304Z",
    "criticalSystem": false,
    "description": "This is a cool project.",
    "id": 4,
    "name": "Project X",
    "namespacePrefix": "4ea35c",
    "primaryClusterName": "silver",
    "prioritySystem": false,
    "updatedAt": "2020-09-10T18:14:13.304Z",
    "userId": 1,
  },
]
`;

exports[`Project-profile event handlers Fetch all profiles should throw 1`] = `"Unable fetch all project profiles"`;

exports[`Project-profile event handlers Fetch all profiles should throw 2`] = `Array []`;

exports[`Project-profile event handlers Fetch single profile should throw 1`] = `"Unable fetch project profile with ID 4"`;

exports[`Project-profile event handlers Fetch single profile should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Project-profile event handlers Unique profile names are derived 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT COUNT(*) FROM profile
          WHERE namespace_prefix = $1;",
      "values": Array [
        "c8c7e6",
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT COUNT(*) FROM profile
          WHERE namespace_prefix = $1;",
      "values": Array [
        "c8c7e6",
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT COUNT(*) FROM profile
          WHERE namespace_prefix = $1;",
      "values": Array [
        "c8c7e6",
      ],
    },
  ],
]
`;

exports[`Project-profile event handlers Unique profile names are derived 2`] = `"c8c7e6"`;
